openmc/data/correlated.py

```diff
@@ -0,0 +1,462 @@
+        return self._breakpoints
+
+    @property
+    def interpolation(self):
+        return self._interpolation
+
+    @property
+    def energy(self):
+        return self._energy
+
+    @property
+    def energy_out(self):
+        return self._energy_out
+
+    @property
+    def mu(self):
+        return self._mu
+
+    @breakpoints.setter
+    def breakpoints(self, breakpoints):
+        cv.check_type('correlated angle-energy breakpoints', breakpoints,
+                      Iterable, Integral)
+        self._breakpoints = breakpoints
+
+    @interpolation.setter
+    def interpolation(self, interpolation):
+        cv.check_type('correlated angle-energy interpolation', interpolation,
+                      Iterable, Integral)
+        self._interpolation = interpolation
+
+    @energy.setter
+    def energy(self, energy):
+        cv.check_type('correlated angle-energy incoming energy', energy,
+                      Iterable, Real)
+        self._energy = energy
+
+    @energy_out.setter
+    def energy_out(self, energy_out):
+        cv.check_type('correlated angle-energy outgoing energy', energy_out,
+                      Iterable, Univariate)
+        self._energy_out = energy_out
+
+    @mu.setter
+    def mu(self, mu):
+        cv.check_iterable_type('correlated angle-energy outgoing cosine',
+                               mu, Univariate, 2, 2)
+        self._mu = mu
+
+    def to_hdf5(self, group):
+        """Write distribution to an HDF5 group
+
+        Parameters
+        ----------
+        group : h5py.Group
+            HDF5 group to write to
+
+        """
+        group.attrs['type'] = np.string_(self._name)
+
+        dset = group.create_dataset('energy', data=self.energy)
+        dset.attrs['interpolation'] = np.vstack((self.breakpoints,
+                                                 self.interpolation))
+
+        # Determine total number of (E,p) pairs and create array
+        n_tuple = sum(len(d.x) for d in self.energy_out)
+        eout = np.empty((5, n_tuple))
+
+        # Make sure all mu data is tabular
+        mu_tabular = []
+        for i, mu_i in enumerate(self.mu):
+            mu_tabular.append([mu_ij if isinstance(mu_ij, (Tabular, Discrete)) else
+                               mu_ij.to_tabular() for mu_ij in mu_i])
+
+        # Determine total number of (mu,p) points and create array
+        n_tuple = sum(sum(len(mu_ij.x) for mu_ij in mu_i)
+                      for mu_i in mu_tabular)
+        mu = np.empty((3, n_tuple))
+
+        # Create array for offsets
+        offsets = np.empty(len(self.energy_out), dtype=int)
+        interpolation = np.empty(len(self.energy_out), dtype=int)
+        n_discrete_lines = np.empty(len(self.energy_out), dtype=int)
+        offset_e = 0
+        offset_mu = 0
+
+        # Populate offsets and eout array
+        for i, d in enumerate(self.energy_out):
+            n = len(d)
+            offsets[i] = offset_e
+
+            if isinstance(d, Mixture):
```

Comment:
There are more `Univariate` types than `Mixture`, `Tabular`, and `Discrete`. Consider checking for those specific types or subclassing `CorrelatedAngleEnergy` to handle these specific cases separately, or you could have unintended side effects when given an unhandled `Univariate`. Conversely, each of those types could "know" how to calculate these values themselves, and you just call the relevant instance method.